components/calendar/calendar.js

```diff
@@ -0,0 +1,474 @@
+		if (this._shownMonth === undefined || !this._shownYear) return [];
+
+		this._dates = [];
+
+		const numDays = this._getNumberOfDaysInMonth(this._shownMonth, this._shownYear);
+
+		// populate first week of month
+		const firstWeek = [];
+		const firstDayOfMonth = new Date(this._shownYear, this._shownMonth, 1).getDay();
+		let numDaysFromLastMonthToShowThisMonth = 0;
+
+		if (firstDayOfMonth !== this._firstDayOfWeek) {
+			const numDaysLastMonth = this._getNumberOfDaysInMonth(this._prevMonth, this._shownYear);
+			numDaysFromLastMonthToShowThisMonth = firstDayOfMonth - this._firstDayOfWeek;
+			if (numDaysFromLastMonthToShowThisMonth < 0) {
+				numDaysFromLastMonthToShowThisMonth += 7;
+			}
+			for (let i = numDaysLastMonth - numDaysFromLastMonthToShowThisMonth + 1; i <= numDaysLastMonth; i++) {
+				firstWeek.push(this._createDateObject(i, true, false));
+			}
+		}
+		for (let j = 1; j <= 7 - numDaysFromLastMonthToShowThisMonth; j++) {
+			firstWeek.push(this._createDateObject(j, false, false));
+		}
+		this._dates.push(firstWeek);
+
+		// remaining weeks
+		let nextMonthDay = 1;
+		let firstDateOfWeek = 7 - numDaysFromLastMonthToShowThisMonth + 1;
+		for (let i = 1; i < Math.ceil((numDaysFromLastMonthToShowThisMonth + numDays) / 7); i++) {
+			const week = [];
+			for (let j = firstDateOfWeek; j < firstDateOfWeek + 7; j++) {
+				let day;
+				if (j >= (numDays + 1)) {
+					day = this._createDateObject(nextMonthDay, false, true);
+					nextMonthDay++;
+				} else {
+					day = this._createDateObject(j, false, false);
+				}
+				week.push(day);
+			}
+			firstDateOfWeek = firstDateOfWeek + 7;
+			this._dates.push(week);
+		}
+
+		return this._dates;
+	}
+
+	_getCalendarDateByDate(date) {
+		return this.shadowRoot.querySelector(`d2l-calendar-date[date="${date.date}"][month="${date.month}"][year="${date.year}"]`);
+	}
+
+	_getNodeAndAddFocus(keyboardTriggeredFocus) {
+		const node = this._getCalendarDateByDate(this._focusDate);
+		if (node) {
+			node.setAttribute('focused', true);
+			if (keyboardTriggeredFocus) {
+				node.focus();
+			}
+		}
+	}
+
+	_getNodeAndRemoveFocus() {
+		const node = this._getCalendarDateByDate(this._focusDate);
+		if (node) {
+			node.removeAttribute('focused');
+		}
+	}
+
+	_getNumberOfDaysInMonth(month, year) {
+		month++; // month index started at 0
+		let days = 31;
+		if (month === 2) {
+			if ((year % 4 === 0) && ((year % 100 !== 0) || (year % 400 === 0))) {
+				days = 29;
+			} else {
+				days = 28;
+			}
+		} else if (month === 4 || month === 6 || month === 9 || month === 11) {
+			days = 30;
+		}
+		return days;
+	}
+
+	_onKeyDown(e) {
+		const rootTarget = e.composedPath()[0];
+		if (rootTarget.tagName !== 'D2L-CALENDAR-DATE') return;
+
+		const keyCodes = {
+			DOWN: 40,
+			END: 35,
+			ENTER: 13,
+			ESCAPE: 27,
+			HOME: 36,
+			LEFT: 37,
+			PAGEUP: 33,
+			PAGEDOWN: 34,
+			SPACE: 32,
+			RIGHT: 39,
+			UP: 38
+		};
+
+		let preventDefault = false;
+
+		switch (e.keyCode) {
```

Comment:
Question: when initially focusing on the  calendar via keyboard, where should focus go, say for focusing forwards and also focusing backwards? If tabbing backwards, should focus go to the date, or should it go to the same place as when focusing forwards?

Reply:
For focusing forwards, focus initially goes to the previous month arrow. I'm not sure about focusing backwards. I think it would make sense to go to the same place as when focusing forwards.